### PR TITLE
Create gradient color picker

### PR DESCRIPTION
**Pull request by sutser**

### CODE REVIEW

File: gradient-color-picker/.travis.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+dist: bionic
```
Comment (tautropfli) [inline gradient-color-picker/.travis.yml]: This should be in the root of the repository

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,288 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (tautropfli) [inline .gitignore]: Use this gitignore https://github.com/messerli-informatik-ag/session/blob/master/.gitignore
And add an entry for `paket.lock`

File: gradient-color-picker/gradient-color-picker.csproj [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <RootNamespace>gradient_color_picker</RootNamespace>
+    <ApplicationIcon />
+    <StartupObject />
+    <Nullable>enable</Nullable>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+  </PropertyGroup>
+  <Import Project="..\.paket\Paket.Restore.targets" />
+</Project>
```
Comment (tautropfli) [inline gradient-color-picker/gradient-color-picker.csproj]: Missing trailing newline

Comment (tautropfli) [reply gradient-color-picker/gradient-color-picker.csproj]:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <RootNamespace>Messerli.GradientColorPicker</RootNamespace>
        <AssemblyName>Messerli.GradientColorPicker</AssemblyName>
    </PropertyGroup>
    <PropertyGroup>
        <TargetFramework>netstandard2.1</TargetFramework>
        <LangVersion>8.0</LangVersion>
        <Nullable>enable</Nullable>
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
        <GenerateDocumentationFile>true</GenerateDocumentationFile>
    </PropertyGroup>
</Project>

File: paket.dependencies [hunk 4]
```diff
@@ -0,0 +1,6 @@
+framework: auto-detect
+
+source https://api.nuget.org/v3/index.json
+source https://www.myget.org/F/natemcmaster/api/v3/index.json
```
Comment (tautropfli) [inline paket.dependencies]: Remove this

File: gradient-color-picker/obj/project.assets.json [hunk 5]
```diff
@@ -0,0 +1,356 @@
+{
```
Comment (tautropfli) [inline gradient-color-picker/obj/project.assets.json]: Remove this file

File: gradient-color-picker/obj/gradient-color-picker.csproj.nuget.g.targets [hunk 6]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8" standalone="no"?>
```
Comment (tautropfli) [inline gradient-color-picker/obj/gradient-color-picker.csproj.nuget.g.targets]: Remove this

File: gradient-color-picker/obj/Debug/netcoreapp3.0/gradient-color-picker.AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+//------------------------------------------------------------------------------
```
Comment (tautropfli) [inline gradient-color-picker/obj/Debug/netcoreapp3.0/gradient-color-picker.AssemblyInfo.cs]: Remove this

File: gradient-color-picker/Program.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (tautropfli) [inline gradient-color-picker/Program.cs]: Remove this

File: gradient-color-picker-test/GradientColorBuilderTest.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+            Assert.Equal(gradientColorProvider.ColorPicker(0), Color.Green);
+        }
+
+        [Fact]
+        public void ErrorsOnSameValueAdded()
+        {
+            const int sameValue = 0;
+            Assert.Throws<InvalidOperationException>(() =>
+                WithColor(new GradientColorByValue(Color.Green, sameValue))
+                    .Add(new GradientColorByValue(Color.Green, sameValue))
```
Comment (tautropfli) [inline gradient-color-picker-test/GradientColorBuilderTest.cs]: Add another test that calls Add() twice with the same value

File: gradient-color-picker-test/GradientColorBuilderTest.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+using static Gradient_color_picker.GradientColorBuilder;
+
+namespace Gradient_color_picker_test
+{
+    public partial class GradientColorBuilderTest
+    {
+        [Fact]
+        public void CreateGradientColorBuilder()
+        {
+            var gradientColorProvider = WithColor(new GradientColorByValue(Color.Red, -10))
```
Comment (tautropfli) [inline gradient-color-picker-test/GradientColorBuilderTest.cs]: Test that the builder works when `Add` is never called.

File: gradient-color-picker-test/GradientColorProviderTest.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        [Fact]
+        public void UsesMaximumPositionRangeOfColor()
+        {
+            Assert.Equal(_gradientColorProvider.ColorPicker(20), Color.Blue);
+        }
+
+        [Fact]
+        public void CalculatesLinearInterpolationForGradientColor()
```
Comment (tautropfli) [inline gradient-color-picker-test/GradientColorProviderTest.cs]: Use a theory

File: gradient-color-picker-test/gradient-color-picker-test.csproj [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <RootNamespace>gradient_color_picker_test</RootNamespace>
+
+    <IsPackable>false</IsPackable>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+      <GenerateDocumentationFile>true</GenerateDocumentationFile>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
```
Comment (tautropfli) [inline gradient-color-picker-test/gradient-color-picker-test.csproj]: Reference dependencies via paket

File: gradient-color-picker-test/gradient-color-picker-test.csproj [hunk 13]
```diff
@@ -0,0 +1,23 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <RootNamespace>gradient_color_picker_test</RootNamespace>
```
Comment (tautropfli) [inline gradient-color-picker-test/gradient-color-picker-test.csproj]: Correct formatting and use c#8 (see comment on other csproj file)

File: gradient-color-picker/GradientColorBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            _gradientColorList = gradientColorList;
+        }
+
+        public static GradientColorBuilder WithColor(GradientColorByValue colorByValue)
+            => new GradientColorBuilder().Add(colorByValue);
+
+        public GradientColorBuilder Add(GradientColorByValue gradientColor)
+        {
+            if (_gradientColorList.ToList().Any(item => item.Value == gradientColor.Value))
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorBuilder.cs]: Is the `.ToList()` required?

File: gradient-color-picker/GradientColorBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                throw new InvalidOperationException();
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorBuilder.cs]: Add a message

File: gradient-color-picker/GradientColorByValue.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Drawing;
+
+namespace Gradient_color_picker
+{
+    public class GradientColorByValue
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorByValue.cs]: Implement equals on this

File: gradient-color-picker/GradientColorProvider.cs [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    public class GradientColorProvider : IGradientColorProvider
+    {
+        private readonly IReadOnlyCollection<GradientColorByValue> _gradientColorList;
+
+        public GradientColorProvider(IImmutableList<GradientColorByValue> gradientColorList)
+        {
+            _gradientColorList = gradientColorList;
+        }
+
+        public Color ColorPicker(int value)
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Rename this to a verb

File: gradient-color-picker/GradientColorProvider.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+            var nextSmallerGradientColorByValue = _gradientColorList.ToList().LastOrDefault(item => item.Value < value);
+            var nextHigherGradientColorByValue = _gradientColorList.ToList().FirstOrDefault(item => item.Value > value);
+
+            return GetColor(nextSmallerGradientColorByValue, nextHigherGradientColorByValue, value);
+        }
+
+        private static Color GetColor(GradientColorByValue? smallerGradientColorByValue, GradientColorByValue? higherGradientColorByValue, int value)
+        {
+            if (smallerGradientColorByValue == null && higherGradientColorByValue == null)
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Use `is` to compare against `null`.

File: gradient-color-picker/LinearInterpolation.cs [hunk 19]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Gradient_color_picker
+{
+    public class LinearInterpolation
```
Comment (tautropfli) [inline gradient-color-picker/LinearInterpolation.cs]: Make this class `static`

File: gradient-color-picker/GradientColorProvider.cs [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        public Color ColorPicker(int value)
+        {
+            if (_gradientColorList.Any(item => item.Value == value))
+            {
+                return _gradientColorList.First(item => item.Value == value).Color;
+            }
+
+            var nextSmallerGradientColorByValue = _gradientColorList.ToList().LastOrDefault(item => item.Value < value);
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: `.LastOrDefault` should work directly on `_gradientColorList`, right?

Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Use a sliding window instead: https://stackoverflow.com/a/3390536/6827589
There are two edge-cases to check: `value <= _gradientColorList.First()` and `value >= _gradientColorList.Last()`

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (sutser) [reply .gitignore]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/8e33974dce0f2f6d242911f49025859603f717c3

File: gradient-color-picker/.travis.yml (hunk 1 above)
Comment (sutser) [reply gradient-color-picker/.travis.yml]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/450200fa218a76ecd34b400f342a35d433ab0c2d

File: gradient-color-picker/obj/Debug/netcoreapp3.0/gradient-color-picker.AssemblyInfo.cs (hunk 7 above)
Comment (sutser) [reply gradient-color-picker/obj/Debug/netcoreapp3.0/gradient-color-picker.AssemblyInfo.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/be66ebf6c7691327421f45b1e8341cb430f5ed9d

File: gradient-color-picker/obj/gradient-color-picker.csproj.nuget.g.targets (hunk 6 above)
Comment (sutser) [reply gradient-color-picker/obj/gradient-color-picker.csproj.nuget.g.targets]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/be66ebf6c7691327421f45b1e8341cb430f5ed9d

File: gradient-color-picker/gradient-color-picker.csproj (hunk 3 above)
Comment (sutser) [reply gradient-color-picker/gradient-color-picker.csproj]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/e7df54d2b8f19b9679a3452616732510c3c2bcfa

File: paket.dependencies (hunk 4 above)
Comment (sutser) [reply paket.dependencies]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/9a878f465907600f307abb3168d1aeafcc120205

File: gradient-color-picker/obj/project.assets.json (hunk 5 above)
Comment (sutser) [reply gradient-color-picker/obj/project.assets.json]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/be66ebf6c7691327421f45b1e8341cb430f5ed9d

File: gradient-color-picker/Program.cs (hunk 8 above)
Comment (sutser) [reply gradient-color-picker/Program.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/9190dd39820e1a14fa943778c345fd3944c82b0e

File: gradient-color-picker-test/GradientColorBuilderTest.cs (hunk 9 above)
Comment (sutser) [reply gradient-color-picker-test/GradientColorBuilderTest.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/9ec7532c3d6e878fed59fd5ad516fe405da205c8

File: gradient-color-picker-test/GradientColorBuilderTest.cs (hunk 10 above)
Comment (sutser) [reply gradient-color-picker-test/GradientColorBuilderTest.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/fc5c50de5ffdb20d4850f84a9a96fd3522ceaffc

File: gradient-color-picker-test/gradient-color-picker-test.csproj (hunk 13 above)
Comment (sutser) [reply gradient-color-picker-test/gradient-color-picker-test.csproj]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/90439ad6f051825389ee2cd63cb60dae3d747556

File: gradient-color-picker/GradientColorBuilder.cs (hunk 14 above)
Comment (sutser) [reply gradient-color-picker/GradientColorBuilder.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/3f577d0484e8502c701ca8e8fc13b75d04e9d325

File: gradient-color-picker/GradientColorBuilder.cs (hunk 15 above)
Comment (sutser) [reply gradient-color-picker/GradientColorBuilder.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/132c4c32bf5a1b4a968cc6b027924989e8b25638

File: gradient-color-picker/GradientColorProvider.cs (hunk 17 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/a139d9736bc96cc32aba8cd7e4bbce33cd564e9c

File: gradient-color-picker/GradientColorProvider.cs (hunk 18 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/aa97dc7baff352fc329c7a33acbfdb29771644c0

File: gradient-color-picker/GradientColorProvider.cs (hunk 17 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/734a6b1237055cd1ffc52fb1cc747ddac28fc69f

Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/a139d9736bc96cc32aba8cd7e4bbce33cd564e9c

File: gradient-color-picker/GradientColorProvider.cs (hunk 18 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/aa97dc7baff352fc329c7a33acbfdb29771644c0

File: gradient-color-picker-test/gradient-color-picker-test.csproj (hunk 12 above)
Comment (sutser) [reply gradient-color-picker-test/gradient-color-picker-test.csproj]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/commit/90439ad6f051825389ee2cd63cb60dae3d747556

File: gradient-color-picker-test/gradient-color-picker-test.csproj (hunk 13 above)
Comment (tautropfli) [reply gradient-color-picker-test/gradient-color-picker-test.csproj]: Could you re-add `<IsPackable>false</IsPackable>`?

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: gradient-color-picker/LinearInterpolation.cs (hunk 19 above)
Comment (sutser) [reply gradient-color-picker/LinearInterpolation.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/8f023219e4c816d38a2c04ff2dfdccf21d6164fe

File: gradient-color-picker-test/gradient-color-picker-test.csproj (hunk 13 above)
Comment (sutser) [reply gradient-color-picker-test/gradient-color-picker-test.csproj]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/6b553f3537100a30111dd7974c64a53b0b8db845

File: gradient-color-picker-test/GradientColorProviderTest.cs (hunk 11 above)
Comment (sutser) [reply gradient-color-picker-test/GradientColorProviderTest.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/6aac651fb1acd140c18ed7d45f57f44f91feed5c

File: gradient-color-picker/GradientColorByValue.cs (hunk 16 above)
Comment (sutser) [reply gradient-color-picker/GradientColorByValue.cs]: Addresses in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/29c4295c165ad80c48a51092e60d8c6037fe1e2b

File: readme.md [hunk 21]
```diff
@@ -0,0 +1,6 @@
+# gradient-color-picker
+
+[![Build Status](https://travis-ci.org/messerli-informatik-ag/test-utility.svg?branch=master)](https://travis-ci.org/messerli-informatik-ag/test-utility)
```
Comment (dngroth) [inline readme.md]: `test-utility` should be `gradient-color-picker`

File: readme.md [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+[![NuGet](https://img.shields.io/nuget/v/Messerli.Test.Utility.svg)](https://www.nuget.org/packages/Messerli.Test.Utility/)
```
Comment (dngroth) [inline readme.md]: `Messerli.Test.Utility` should be `Messerli.GradientColorPicker`

File: gradient-color-picker/GradientColorProvider.cs (hunk 20 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/72a6125219c08d30e12ff2187b3996c942993150

Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Falls away due to the change of https://github.com/messerli-informatik-ag/gradient-color-picker/commit/72a6125219c08d30e12ff2187b3996c942993150

File: gradient-color-picker/GradientColorProvider.cs [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Drawing;
+using System.Linq;
+
+namespace Messerli.GradientColorPicker
+{
+    public class GradientColorProvider : IGradientColorProvider
+    {
+        private readonly IReadOnlyCollection<GradientColorByValue> _gradientColorList;
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]:
```suggestion
        private readonly IImmutableList<GradientColorByValue> _gradientColorList;
```

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: gradient-color-picker/GradientColorProvider.cs [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+        public GradientColorProvider(IImmutableList<GradientColorByValue> gradientColorList)
+        {
+            _gradientColorList = gradientColorList.OrderBy(item => item.Value).ToImmutableList();
+        }
+
+        public Color PickColor(int value)
+        {
+            if (!_gradientColorList.Any())
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Put this into a function

File: gradient-color-picker/GradientColorProvider.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                throw new InvalidOperationException("No gradient color defined");
+            }
+
+            if (value <= _gradientColorList.First().Value)
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Put `_gradientColorList.First()` into a variable

File: gradient-color-picker/GradientColorProvider.cs [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                return _gradientColorList.First().Color;
+            }
+
+            if (value >= _gradientColorList.Last().Value)
```
Comment (tautropfli) [inline gradient-color-picker/GradientColorProvider.cs]: Put `_gradientColorList.Last()` into a variable.

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: gradient-color-picker-test/GradientColorBuilderTest.cs (hunk 9 above)
Comment (sutser) [reply gradient-color-picker-test/GradientColorBuilderTest.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/f294c12c1116d67f1f1bb49a8fa1cb1785a8c806

File: gradient-color-picker/GradientColorProvider.cs (hunk 24 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/070eb06dbe23669a7311f7a81e1ad4f94db22c2c

File: gradient-color-picker/GradientColorProvider.cs (hunk 26 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/be8d2e6f03d9373895cccf4f0b071f2c7166eb65

File: gradient-color-picker/GradientColorProvider.cs (hunk 25 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/c532a0b0c0c2f76cf2ab557909d30ae047dc3a9e

File: gradient-color-picker/GradientColorProvider.cs (hunk 24 above)
Comment (sutser) [reply gradient-color-picker/GradientColorProvider.cs]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/070eb06dbe23669a7311f7a81e1ad4f94db22c2c

**Review by tautropfli — COMMENTED** [review]
Looks good to me so far, except for the build error and the readme changes suggested by @dngroth.

File: readme.md (hunk 21 above)
Comment (sutser) [reply readme.md]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/48601bb0f8ee4665a21f47cbbd8a18f94cbbcff6

File: readme.md (hunk 22 above)
Comment (sutser) [reply readme.md]: Addressed in: https://github.com/messerli-informatik-ag/gradient-color-picker/pull/1/commits/48601bb0f8ee4665a21f47cbbd8a18f94cbbcff6

**Review by tautropfli — APPROVED** [review]

**Pull request merged**